### ISSUE TITLE
How to bulk insert or update with custom unique key? I am getting "Column 'Id' does not allow DBNull.Value" error.

### ISSUE THREAD

**Issue by Ninjanaut** [borisdj/EFCore.BulkExtensions#626]
Hello, 

I am getting following error

> System.InvalidOperationException: 'Column 'Id' does not allow DBNull.Value.'

My SQL Server table looks like
```sql
create table [Foo].[Order]
(
	[Id] int identity not null,

	Col1  nvarchar(255),
	Col2  nvarchar(255),
	Col3  nvarchar(255),

	[ImportedOn] datetime not null,

	constraint PK_Foo_Order primary key (Id),
	constraint AK_Foo_Order unique (Col1 Col2)
)
```
My entity looks like

```csharp
[Table("Order", Schema = "Foo")]
public class Order : IEquatable<Order>
{
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public int? Id { get; set; }

    public string Col1 { get; set; } // Private set originally, I am not sure if it plays a role.
    public string Col2 { get; set; } // Part of Unique key in the database
    public string Col3 { get; set; } // Part of Unique key in the database

    public DateTime ImportedOn { get; set; }

    public Order() { } // BulkExtension does not know how to read private ctor, so we have to break the encapsulation.

    public Order(string col1, string col2, string col3)
    {
        Col1 = col1;
        Col2 = col2;
        Col3 = col3;

        ImportedOn = DateTime.Now;
    }

    public override bool Equals(object obj)
    {
        return Equals(obj as Order);
    }

    public bool Equals(Order other)
    {
        return  other != null && 
                Col1 == other.Col1 &&
                Col2 == other.Col2;
    }

    public override int GetHashCode()
    {
        return HashCode.Combine(Col1, Col2);
    }
}
```

I would like to insert data whose unique key is not in the database and update the rows where the unique key matches. 

I admit, however, that the documentation regarding the properties like
* PropertiesToInclude
* PropertiesToIncludeOnCompare
* PropertiesToIncludeOnUpdate

is not clear to me and is rather confusing.

However I tried following command

```csharp
using var transaction = _database.Database.BeginTransaction();

_database.BulkInsertOrUpdate(orders, new BulkConfig() { 
     UpdateByProperties = new List<string>() { 
          // The unique key definition here
          nameof(Order.Col1),
          nameof(Order.Col2),
      } 
  }); 

transaction.Commit();
```

which fails.

I also tried 

```csharp
using var transaction = _database.Database.BeginTransaction();

_database.Orders.Where(order => orders.Contains(order)).BatchDelete();
_database.BulkInsert(orders.Distinct().ToList());

transaction.Commit();
```

but I get the unique key error when inserting.

Can you please help me how to proceed with this simple operation? Am I using the right method?

Best regards,
Ninjanut

**Comment by borisdj** [comment]
In entity model instead of:
`public int? Id { get; set; }`
define it as NonNullable:
`public int? Id { get; set; }`

**Comment by Ninjanaut** [comment]
Thank you @borisdj for a quick response. I changed primary key to **NonNullable**

```csharp
[DatabaseGenerated(DatabaseGeneratedOption.Identity)]
public int Id { get; set; }
```

and also add **PropertiesToExclude** property

```csharp
using var transaction = _database.Database.BeginTransaction();

_database.BulkInsertOrUpdate(orders, new BulkConfig() { 
     UpdateByProperties = new List<string>() { 
          // The unique key definition here
          nameof(Order.Col1),
          nameof(Order.Col2),
      },
      PropertiesToExclude = { nameof(Order.Id) } 
  }); 

transaction.Commit();
```

But now I have another error

> System.NullReferenceException: 'Object reference not set to an instance of an object.' EFCore.BulkExtensions.BulkConfig.PropertiesToExclude.get returned null.

what does it mean, what am I missing?

**Comment by borisdj** [comment]
'Id' should not be Excluded

**Comment by Ninjanaut** [comment]
Thank you @borisdj, I finally found that the error was in the underlying data where I had unexpected duplications. It works now! :-) The code works both way, with or without
```csharp 
PropertiesToExclude = { nameof(Order.Id) }
```

What is confusing me is an instruction from the method note, which clearly says that the **Id** column should be excluded.

![BORISDJ_EFCORE_ERROR](https://user-images.githubusercontent.com/44192515/132061832-d4734823-5510-4093-9676-137bf06d8e19.png)
